### DISCUSSION TITLE
Could xsbug-log work with mcrun and also _not_ use XSBUG_HOST/XSBUG_PORT?

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1117] [General]
I'm trying to use xsbug-log with mods and it's not really working for me. The issues I've encountered:
- mcrun doesn't launch xsbug-log (seems to ignore -l)
- xsbug-log uses XSBUG_HOST/XSBUG_PORT, which is not really cli friendly IMHO

On the second point, if I ask for debug to show on stdout then I kind'a expect it to show up there and not in some other terminal that happens to have some xsbug-log running, which is what happens now. I remember there was some discussion whether xsbug-log should or should not kill any running xsbug and I see that it prints a message. Why not simply use a random port between serial2xsbug and xsbug-log which would (a) allow as many concurrent uses as one has terminals and (b) always show the output in the terminal where mcconfig/mcrun are run, and (c) avoid any conflict with xsbug?

**Update:**
Looking at the source of xsbug-log I notice that it's fixed on localhost:5002 and that it launches serial2xsbug, I suspect this means that if I have XSBUG_* environment variables set to different values that xsbug-log and serial2xsbug won't actually communicate...

**Comment by tve** [comment]
Looks like it's easy to change the xsbug-log behavior to be as I would like. I'm open to suggestions on how to configure this (right now there's a `scan_ports` constant) and could write a PR. I also suspect that I want `autoexit=true` but am not exactly sure what triggers the exit.

diff:
```
--- a/tools/xsbug-log/xsbug-log.js
+++ b/tools/xsbug-log/xsbug-log.js
@@ -70,24 +70,34 @@ class LogMachine extends Machine {
        }
 }
 
-const portIn = 5002;
+let portIn = 5002;
 let connections = 0;
 let autoexit = false;
-
-let probe = net.connect({
-       port: portIn,
-       host: "127.0.0.1"
-});
-probe.setEncoding("utf8");
-probe.on('ready', data => {
-       console.log(`#xsbug-log: Debugger detected running on port ${portIn}. Please quit xsbug and try again.`);
-       process.exit(-1);
-});
-probe.on('error', error => {
-       // no debugger detected. continue.
-       probe.destroy();
-       launch();
-});
+const scan_ports = true;
+function try_port() {
+       console.log("#xsbug-log: trying port " + portIn);
+       let probe = net.connect({
+               port: portIn,
+               host: "127.0.0.1"
+       });
+       probe.setEncoding("utf8");
+       probe.on('ready', data => {
+               if (scan_ports) {
+                       probe.destroy();
+                       portIn++;
+                       try_port();
+               } else {
+                       console.log(`#xsbug-log: Debugger detected running on port ${portIn}. Please quit xsbug and try again.`);
+                       process.exit(-1);
+               }
+       });
+       probe.on('error', error => {
+               // found free port, continue.
+               probe.destroy();
+               launch();
+       });
+}
+try_port();
 
 function launch() {
        const server = net.createServer(target => { 
@@ -116,5 +126,7 @@ function launch() {
        for (let i = 3; i < process.argv.length; i++) 
                command += ` ${process.argv[i]}`;
 
+       process.env.XSBUG_HOST = '127.0.0.1';
+       process.env.XSBUG_PORT = portIn;
        exec(command);
 }
```

**Comment by phoddie** [comment]
> xsbug-log uses XSBUG_HOST/XSBUG_PORT

There is no reference to either `XSBUG_HOST` or `XSBUG_PORT` in [xsbug-log.js](https://github.com/Moddable-OpenSource/moddable/blob/public/tools/xsbug-log/xsbug-log.js). Is it possible your repository is not up-to-date?

> mcrun doesn't launch xsbug-log (seems to ignore -l)

Support launching xsbug-log from mcrun not yet in place.

Both the mcconfig and mcrun tools set `XSBUG_LOG` in the generated makefile. What remains to update the [mcrun makefile templates](https://github.com/Moddable-OpenSource/moddable/tree/public/tools/mcrun) to use it.

**Reply by tve** [reply]
You are correct that xsbug-log itself doesn't use XSBUG_*. However, if these env vars are set they are passed through to serial2xsbug and the latter then potentially launches on a port different from xsbug-log. But in any case, I believe something along the lines of the patch I posted makes it cleaner and completely independent of any running xsbug.
(Also, the patch allows log messages printed by serial2xsbug to be seen...)

**Reply by phoddie** [reply]
There was quite a bit of discussion about this behavior a while back. That has settled on the current behavior. I would only reconsider that with a strong motivation and input from those involved.

>  if I ask for debug to show on stdout then I kind'a expect it to show up there and not in some other terminal that happens to have some xsbug-log running, which is what happens now. 

When launching xsbug-log with another debugger instance active, xsbug-log lets you know that.

```
#xsbug-log: Debugger detected running on port 5002. Please quit xsbug and try again.
```

**Reply by tve** [reply]
Ah, I finally understand what happened. I launched `mcconfig -l` in one terminal. Then I launched `mcrun -l` in another terminal. Since `mcrun` silently ignores `-l` (I was not aware of that at the time) the debug output showed up in the `mcconfig` terminal. I therefore concluded that `xsbug-log` uses an existing debugger as I was unaware that `xsbug-log` was not actually invoked. That lead me to write the above patch to try and remedy the situation.

**Reply by phoddie** [reply]
Great, so that is resolved.

Are you motivated to take a run at the makefile updates to support xsbug-log in mcrun?

**Reply by tve** [reply]
Once https://github.com/Moddable-OpenSource/moddable/issues/1115 is fixed I can see whether I can grok mcrun's makefiles...
